### DISCUSSION TITLE
VectorFitting Network Synthesis References

### DISCUSSION THREAD

**Discussion by fnordpole** [scikit-rf/scikit-rf#1172] [Q&A] [answered]
Hi,

in vectorFitting.py in the method write_spice_subcircuit_s, a network is synthesized. In the code, the paper from G. Antonini is referenced. This paper describes only the realizations of complex pole pairs and real poles using RL or RLCG circuits.

I am missing here a reference to the actual synthesis approach. There is quite a bit of literature about this topic, for example
"Equivalent Circuit Synthesis of Multiport S Parameters in Pole–Residue Form", C. Chou and José E. Schutt-Ainé but I could not find the exact approach implemented here in scikit-rf.

If the used approach is new and unpublished, a bit of documentation would be great. Otherwise, a reference to the respective paper would be appreciated.

Thomas

**Comment by Vinc0110** [comment, marked as answer]
Hi Thomas,
the synthesis was a very tricky part and it took me days to figure it out. I would not call it "new", but I could not find any references describing the process in all the details required. In terms of documentation of "my" method, I wrote several sections in the tutorial:

https://scikit-rf.readthedocs.io/en/latest/tutorials/VectorFitting.html#Equivalent-Circuit-of-a-Vector-Fitted-N-Port

One section above you will find the individual equivalent circuits from Antonini's paper, which I adopted and combined into the full equivalent circuit.

I hope this documentation does answer your question, but feel free to ask for more details. 

By the way, I did my best to also comment the code itself. Have a look, it might also explain things better:
https://github.com/scikit-rf/scikit-rf/blob/master/skrf/vectorFitting.py#L2260

Regards
Vincent

**Comment by fnordpole** [comment]
Hi Vincent,

thank you for your reply! I indeed haven't seen your approach of having those _inv inverted F sources for the negative residues/poles in the literature. There seem to be different ways of converting the negative values into positive-only realizations.

A little bit of background why I am investigating this: I simulated the s-parameterts of a differential transmission line with two conductors in spectre using the built-in bbspice method. The line has 4 ports, p1 to p4.
The ports map to the conductors in this way:  p1 -> p3 (conductor p) and p2 -> p4 (conductor n). 

I then created a vector-fitted spice netlist using scikit-rf and simulated it in spectre and compared the results to the spectre built-in results. They are for high frequencies pretty much identical. 

Then I did a DC experiment: I connected a 1mA DC current source to p1 and connected a 1kOhm resistor to p3. I expect the full 1 mA current to flow through this 1kOhm resistor. Using the spectre built-in approach, I actually see 1.000 mA. Using the vector-fitted model, I only get 0.984 mA. So about 17e-6 amperes are "missing". 

I want to find out where in the vector-fitted model those 17 microamperes "get lost". 

In the code you mentioned, I saw those comments:

                # CCVS and VCVS driving the transfer admittances with a = V/2/sqrt(Z0) + I/2*sqrt(Z0)
                # In
                f.write(f'H{n + 1} nt{n + 1} nts{n + 1} V{n + 1} {np.real(self.network.z0[0, n])}\n')
                # Vn
                f.write(f'E{n + 1} nts{n + 1} 0 p{n + 1} 0 {1}\n')

You write about "driving the transfer admittances with a=..." and you have some square roots there. But for some reason, in the code I could not find those square roots. Is it a bug or did they cancel out somehow? So the driving of the transfer admittances is actually implemented with gain=1 for the E source and "gain"=Z0 for the H source. 

So my hypothesis was that maybe one of those square roots could be missing, which could then introduce a very little error that I would not see in the high-frequency S-parameters, but maybe at DC.  I did all kinds of experiments with those square roots but none of them fixed the DC problem. So maybe it's all right and my hypothesis is wrong.

Note: I have right now no clue to what extent the actual vector-fit influences the DC behavior of a synthesized circuit. Maybe in spectre they implemented some extra algorithm to "fix" the DC behavior that is not implemented in scikit-rf. If this is true, then the circuit-synthesis code is probably the wrong place to look in order to fix the DC behavior and a fix would rather have to be implemented in the fitting or post-fitting.

Thomas

**Comment by Vinc0110** [comment]
This is very interesting! I don't thing I ever tested the dc solution.

I'm skeptical about your hypothesis. Keep in mind that $a_i$ is the incident power wave at port $i$, it is not the gain of the controlled sources. The implementation does match the schematic of the transfer network in the tutorial. But I would have to dive back into the math to prove it properly...

Could you provide more details on your data? Do your simulated S-parameters include the dc solution? A comparison of the S-parameter samples and the vector-fitted model would be interesting to see. Can you share a plot from [`plot_s_db()`](https://scikit-rf.readthedocs.io/en/latest/api/generated/skrf.vectorFitting.VectorFitting.plot_s_db.html)? I think there you will find the reason for the "lost" current: it is probably already inaccurate in the vector-fitted model and only gets copied into the SPICE netlist.

It would not be surprising if there is an error between the data and the model at the dc point. Keep in mind that the vector fitting algorithm minimizes the global rms error between the data and the model. It does not "care" about the local error at a specific frequency. For that reason, you will likely get different behaviors at the dc point from different pole configurations (number and type of poles). More (real) poles might help.

Do you use regular vector fitting ([`vector_fit()`](https://scikit-rf.readthedocs.io/en/latest/api/generated/skrf.vectorFitting.VectorFitting.vector_fit.html)) or the automatic one ([`auto_fit()`](https://scikit-rf.readthedocs.io/en/latest/api/generated/skrf.vectorFitting.VectorFitting.auto_fit.html))?

**Comment by fnordpole** [comment]
I think you are right, the cause appears to be in the vector fitted model itself.

I used autofit to create the model.

I plotted the vector-fit vs. the frequency samples using plot_s_db as you suggested and indeed at DC, I see about -0.02 dB and not 0.00 dB. 

Besides that, the fit and the samples seem to match pretty good.

My data (s4p) contains the DC point, but I am not sure how good the quality of this data point is. It's still a bit interesting how spectre manages to find an exact 0.00 dB DC solution (so that no current "gets lost") while simulating this transmission line.

In case someone is interested in more details on the DC topic, I found two interesting papers related to it:

D. Deschrijver and T. Dhaene, "DC-Preserving Passivity Enforcement for  S -Parameter Based Macromodels," in IEEE Transactions on Microwave Theory and Techniques, vol. 58, no. 4, pp. 923-928, April 2010, doi: 10.1109/TMTT.2010.2042556

I looked into it only shortly but it seems that the passivity-enforce algorithm can affect the DC point. I then compared the scikit-rf vectorFitted model before and after passivity enforcement and I see indeed a difference at DC that appears after the enforce.

P. Liu, W. Thiel, X. Xin, K. Zhu and E. Bracken, "A Novel Causal Method to Blend the DC and AC Solution over the Entire Frequency Band," 2024 IEEE/MTT-S International Microwave Symposium - IMS 2024, Washington, DC, USA, 2024, pp. 470-472, doi: 10.1109/IMS40175.2024.10600406.

The second paper deals with an algorithm to improve DC point of the initial Sparameters obtained from field simulators. As it's pretty new, I guess that the proposed algorithm was not used in the creation of my data that I got out of the field simulator. Maybe with a better preprocess of the initial data it will be easier to obtain an accurate DC solution.

I attached the first two data points of my sparameter data and the samples vs fit plot and two plots zoomed-in at DC with and without passivity enforcement.

Original Data, DC Point and first non-DC point:
![data-s4p-first-two-samples](https://github.com/user-attachments/assets/2f0262d6-3c12-4726-99e2-1d9cd2cc1e7b)

Complete Fit:
![SamplesVsFit](https://github.com/user-attachments/assets/6f6cb019-c164-4efb-b517-4bc78c82b64b)

Zoomed-In at DC without passivity enforce:
![DC-Without-Passivity-Enforce](https://github.com/user-attachments/assets/6f53f7e6-90e1-4899-a7f4-9b52ac959bb7)

Zoomed-In at DC with passivity enforce:
![DC-With-Passivity-Enforce](https://github.com/user-attachments/assets/9f383f5c-67be-45e7-a75f-5b367b0553f5)

**Comment by Vinc0110** [comment]
I'm glad you found the cause. Is it really a problem or have you just been wondering about  it?

Feel free to create an issue with a feature request, for example of the proposed algorithm for passivity enforcement with dc point preservation. Not sure if I will find time to do it any time soon, but then at least we have it on the to-do list.

Regarding the spectre simulation, maybe they ignore the s parameter model for the DC solution and use a different one only for the DC point...

**Reply by fnordpole** [reply]
For me the accurate DC is indeed important, but I guess for the majority of users it's not a big topic.
It seems also that the DC problem can't be fixed only by implementing the DC preserving passivation algorithm from the paper. 

Thank you for offering to put it on the feature request list, but right now I would not do it. I plan to investigate the topic more and once I have a working prototype I can hopefully make a request to add some extra code to scikit-rf vectorFitting or whatever comes out of the investigation.

In the process so far, I took write_spice_subcircuit_s and adapted it for the spectre simulator, creating write_spectre_subcircuit_s.
If you find it useful you can merge it into the existing vectorFitting class. I am also planning to derive a version for the Xyce simulator.
```python
def write_spectre_subcircuit_s(self, file: str, fitted_model_name: str = "s_equivalent") -> None:
        """
        Creates an equivalent N-port Spectre subcircuit based on its vector fitted S parameter responses.

        Parameters
        ----------
        file : str
            Path and filename including file extension (usually .sp) for the SPICE subcircuit file.
        fitted_model_name: str
            Name of the resulting model, default "s_equivalent"

        Returns
        -------
        None

        Notes
        -----
        In the Spectre subcircuit, all ports will share a common reference node (global Spectre ground on node 0). The
        equivalent circuit uses linear dependent current sources on all ports, which are controlled by the currents
        through equivalent admittances modelling the parameters from a vector fit. This approach is based on [#]_.

        Examples
        --------
        Load and fit the `Network`, then export the equivalent Spectre subcircuit:

        >>> nw_3port = skrf.Network('my3port.s3p')
        >>> vf = skrf.VectorFitting(nw_3port)
        >>> vf.vector_fit(n_poles_real=1, n_poles_cmplx=4)
        >>> vf.write_spice_subcircuit_s('/my3port_model.sp')

        References
        ----------
        .. [#] G. Antonini, "SPICE Equivalent Circuits of Frequency-Domain Responses", IEEE Transactions on
            Electromagnetic Compatibility, vol. 45, no. 3, pp. 502-512, August 2003,
            DOI: https://doi.org/10.1109/TEMC.2003.815528
        """

        # list of subcircuits for the equivalent admittances
        subcircuits = []

        # provides a unique Spectre subcircuit identifier (X1, X2, X3, ...)
        def get_new_subckt_identifier():
            subcircuits.append(f'x{len(subcircuits) + 1}')
            return subcircuits[-1]

        with open(file, 'w') as f:
            # write title line
            f.write('// EQUIVALENT CIRCUIT FOR VECTOR FITTED S-MATRIX\n')
            f.write('// Created using scikit-rf vectorFitting.py\n')
            f.write('//\n')

            # define the complete equivalent circuit as a subcircuit with one input node per port
            # those port nodes are labeled p1, p2, p3, ...
            # all ports share a common node for ground reference (node 0)
            str_input_nodes = ''
            for n in range(self.network.nports):
                str_input_nodes += f'p{n + 1} '

            f.write(f'subckt {fitted_model_name} ({str_input_nodes})\n')

            for n in range(self.network.nports):
                f.write('//\n')
                f.write(f'// port {n + 1}\n')
                # add port reference impedance z0 (has to be resistive, no imaginary part)
                f.write(f'r{n + 1} (a{n + 1} 0) resistor r={np.real(self.network.z0[0, n])}\n')

                # add dummy voltage sources (V=0) to measure the input current
                f.write(f'v{n + 1} (p{n + 1} a{n + 1}) vsource type=dc dc=0\n')

                # CCVS and VCVS driving the transfer admittances with a = V/2/sqrt(Z0) + I/2*sqrt(Z0)
                # In
                f.write(f'h{n + 1} (nt{n + 1} nts{n + 1}) ccvs probe=v{n + 1} rm={np.real(self.network.z0[0, n])}\n')
                # Vn
                f.write(f'e{n + 1} (nts{n + 1} 0 p{n + 1} 0) vcvs gain=1\n')

                for j in range(self.network.nports):
                    f.write(f'// transfer network for s{n + 1}{j + 1}\n')

                    # stacking order in VectorFitting class variables:
                    # s11, s12, s13, ..., s21, s22, s23, ...
                    i_response = n * self.network.nports + j

                    # add CCCS to generate the scattered current I_nj at port n
                    # control current is measured by the dummy voltage source at the transfer network Y_nj
                    # the scattered current is injected into the port (source positive connected to ground)
                    f.write(f'f{n + 1}{j + 1} (0 a{n + 1}) cccs probe=v{n + 1}{j + 1} gain='
                            f'{1 / np.real(self.network.z0[0, n])}\n')
                    f.write(f'f{n + 1}{j + 1}_inv (a{n + 1} 0) cccs probe=v{n + 1}{j + 1}_inv gain='
                            f'{1 / np.real(self.network.z0[0, n])}\n')

                    # add dummy voltage source (V=0) in series with Y_nj to measure current through transfer admittance
                    f.write(f'v{n + 1}{j + 1} (nt{j + 1} nt{n + 1}{j + 1}) vsource type=dc dc=0\n')
                    f.write(f'v{n + 1}{j + 1}_inv (nt{j + 1} nt{n + 1}{j + 1}_inv) vsource type=dc dc=0\n')

                    # add corresponding transfer admittance Y_nj, which is modulating the control current
                    # the transfer admittance is a parallel circuit (sum) of individual admittances
                    f.write(f'// transfer admittances for S{n + 1}{j + 1}\n')

                    # start with proportional and constant term of the model
                    # H(s) = d + s * e  model
                    # Y(s) = G + s * C  equivalent admittance
                    g = self.constant_coeff[i_response]
                    c = self.proportional_coeff[i_response]

                    # add R for constant term
                    if g < 0:
                        f.write(f'r{n + 1}{j + 1} (nt{n + 1}{j + 1}_inv 0) resistor r={np.abs(1 / g)}\n')
                    elif g > 0:
                        f.write(f'r{n + 1}{j + 1} (nt{n + 1}{j + 1} 0) resistor r={1 / g}\n')

                    # add C for proportional term
                    if c < 0:
                        f.write(f'c{n + 1}{j + 1} (nt{n + 1}{j + 1}_inv 0) capacitor c={np.abs(c)}\n')
                    elif c > 0:
                        f.write(f'c{n + 1}{j + 1} (nt{n + 1}{j + 1} 0) capacitor c={c}\n')

                    # add pairs of poles and residues
                    for i_pole in range(len(self.poles)):
                        pole = self.poles[i_pole]
                        residue = self.residues[i_response, i_pole]
                        node = get_new_subckt_identifier() + f' (nt{n + 1}{j + 1}'

                        if np.real(residue) < 0.0:
                            # multiplication with -1 required, otherwise the values for RLC would be negative
                            # this gets compensated by inverting the transfer current direction for this subcircuit
                            residue = -1 * residue
                            node += '_inv'

                        if np.imag(pole) == 0.0:
                            # real pole; add rl_admittance
                            l = 1 / np.real(residue)
                            r = -1 * np.real(pole) / np.real(residue)
                            f.write(node + f' 0) rl_admittance res={r} ind={l}\n')
                        else:
                            # complex pole of a conjugate pair; add rcl_vccs_admittance
                            l = 1 / (2 * np.real(residue))
                            b = -2 * (np.real(residue) * np.real(pole) + np.imag(residue) * np.imag(pole))
                            r = -1 * np.real(pole) / np.real(residue)
                            c = 2 * np.real(residue) / (np.abs(pole) ** 2)
                            gm_add = b * l * c
                            f.write(node + f' 0) rcl_vccs_admittance res={r} cap={c} '
                                           f'ind={l} gm={gm_add}\n')

            f.write('ends s_equivalent\n')

            f.write('//\n')

            # subcircuit for an active RCL+VCCS equivalent admittance Y(s) of a complex-conjugate pole-residue pair H(s)
            # Residue: c = c' + j * c"
            # Pole: p = p' + j * p"
            # H(s)  = c / (s - p) + conj(c) / (s - conj(p))
            #       = (2 * c' * s - 2 * (c'p' + c"p")) / (s ** 2 - 2 * p' * s + |p| ** 2)
            # Y(S)  = (1 / L * s + b) / (s ** 2 + R / L * s + 1 / (L * C))
            f.write('subckt rcl_vccs_admittance (n_pos n_neg)\nparameters res=1e3 cap=1e-9 ind=100e-12 gm=1e-3\n')
            f.write('l1 (n_pos 1) inductor l=ind\n')
            f.write('c1 (1 2) capacitor c=cap\n')
            f.write('r1 (2 n_neg) resistor r=res\n')
            f.write('g1 (n_pos n_neg 1 2) vccs gm=gm\n')
            f.write('ends rcl_vccs_admittance\n')

            f.write('//\n')

            # subcircuit for a passive RL equivalent admittance Y(s) of a real pole-residue pair H(s)
            # H(s) = c / (s - p)
            # Y(s) = 1 / L / (s + s * R / L)
            f.write('subckt rl_admittance (n_pos n_neg)\nparameters res=1e3 ind=100e-12\n')
            f.write('l1 (n_pos 1) inductor l=ind\n')
            f.write('r1 (1 n_neg) resistor r=res\n')
            f.write('ends rl_admittance\n')

```

**Reply by Vinc0110** [reply]
Ok. Thanks for the information. Let me know if I can help.

On first glance it's not obvious for me what the difference from your spectre netlist is compared to the SPICE netlist. The code looks very similar. Does spectre need a special format for the netlist?

**Reply by Vinc0110** [reply]
I found some more information on dc errors and correction methods:
https://www.pscad.com/webhelp-v502-ol/EMTDC/Transmission_Lines/DC_Correction.htm

**Reply by fnordpole** [reply]
The formats for spice, spectre and xyce (and also other simulators I guess) are very similar but unfortunately slightly different. For example the instances in spectre need to have their pin lists in parentheses and the type of device (for example resistor) is not determined by the first letter of the instance name (for example R0) but specified as an extra token "resistor", while the instance name can start with any letter.

Also the subcircuit definition and its parameters is a little different. The gm parameter of the G sources can also be negative in spectre, so I could remove the extra subcircuit parameter "mult" and avoid the expression gm*mult and replace it direcly with the (signed) gm. 

Otherwise I removed the matplotlib EngFormatter calls. Not all the SI prefixes are supported in all the simulators so I found it cleaner to not use them at all, especially in auto-generated netlists and stick to the default exponential/standard number notation instead.

Thank you for looking this up! It looks interesting indeed, I had no time to check it out in detail yet but I will do so soon.

**Comment by fnordpole** [comment]
For those who are interested to know more about the topic of circuit synthesis related to rational function models:

Recently I did a pretty in depth literature search and I can highly recommend these two papers, especially the first one. I guess the method currently impemented in scikit-rf vectorFitting is very similar or even the same as described in (1) Section 2.3 (Circuit synthesis from Spole-residue-form). 

(1) "Circuit synthesis techniques of rational models of electromagnetic systems: A tutorial paper", Romano et al., 2018

and the paper I mentioned already in my first post, 

(2) "Equivalent Circuit Synthesis of Multiport S Parameters in Pole–Residue Form", C. Chou and José E. Schutt-Ainé, 2021

(2) could be especially interesting for scikit-rf for multi-SISO-fitted models because the paper shows that the circuit topology created by using the proposed method are much simpler than the "conventional approaches" and simulation times of those synthesized circuits are reported to be up to 2x faster on most simulators. I guess right now the vectorFitting code in scikit-rf does not support multi-SISO fitting and uses the MIMO fitting approach, but maybe this would be a worthwhile extension for the future.

**Comment by Vinc0110** [comment]
That sounds like an interesting read, but I can't access the papers, unfortunately... Maybe I'll find a way later.

I'm wondering what you mean by "multi-SISO". Is it referring to multiple independent (not coupled) two-port circuits packed into a single SNP file? If that was the case, it would not make much sense to me to attempt to fit them all at once. The fitting itself would be unnecessarily wasteful. Instead of using another optimized synthesis approach, one could just split the file first and then fit and synthesize the individual parts, right? But maybe I totally misunderstand the issue.

**Reply by fnordpole** [reply]
I learned about Multi-SISO vs MIMO fitting in the C. Chou and José E. Schutt-Ainé, 2021 paper. 

MIMO means all Sparameters Sij are fitted using a common set of poles.
Multi-SISO means that for each Sij that is fitted, a set of poles is used.

The paper mentions that Multi-SISO can be advantageous if the Sij have very different poles. Using MIMO, a large number of poles would be used to cover all the variety in all the Sij. But in the fitting process for the individual Sij, this can lead to overfitting because there are way more poles available than the individual Sij require.

Additionally, the paper mentions that Multi-SISO fitting can be easily parallelized and is thus pretty fast because all the Sij use their own set of poles. 

They also write, that Multi-SISO is not necessarily better than MIMO. It depends on the type of data but according to the paper, some commercial tools offer to select one of those methods.

I hope you find a way to get those papers! They are indeed a good read.

**Comment by Vinc0110** [comment]
That sounds reasonable, but I don't understand why those S_ij responses should be packed into a single file in the first place if they are indeed independent (what they apparently call multi-SISO). Of course, as the size of the NxN scattering matrix grows with the square of N, it will be much easier and faster to fit and synthesize each response individually. This will also work in parallel ([Embarrassingly parallel](https://en.wikipedia.org/wiki/Embarrassingly_parallel)).

My point is, I don't think this requires any additional optimized multi-SISO algorithm. You can just extract the individual responses, vector fit them, and synthesize a circuit model for each of them. Of course, they can have (will have) individual sets of poles and the models will be much smaller compared to the whole multi-SISO model. The same works for the circuit simulator, where you can import and connect each model individually.

**Comment by fnordpole** [comment]
I think you are right. The paper has no description on how multi-SISO would be implemented but what you write makes complete sense to me.

Two points that come to my mind that might make it useful still:

Maybe the motivation could be more of a practical one: If you get a sNp with a larger N from an EM simulator and you want to use multi-SISO, you could just tell that to the VF algorithm to treat them separately and use the embarrassingly parallel approach internally. So you would save yourself from exporting a large number of sNp files and feeding them into the VF algorithm individually and then combining all the equivalent circuits.

Another motivation might be the overhead of having N*N port networks instead of N and also N*N transfer networks instead of N. Of course they would all be very simple because they would only deal with a single Sij. But They also have those current-sensor-dc-sources and controlled sources each. This overhead might increase the number of nodes/components in the overall circuit.

**Comment by fnordpole** [comment]
By the way, meanwhile I read those papers entirely and I reviewed your synthesis code in scikit-rf. I think it's too much to post all the details here but for what it is worth, I can confirm that your implementation of the circuit synthesizer code is correct.

The basic idea of having a "Transfer Network" and "Port Networks" is indeed the one presented in C. Chou and José E. Schutt-Ainé, 2021. They call the "Transfer Network" an "Intermediate Stage" and the "Port Network" they call "Port Section".

The method that you implemented to avoid negative R, L and C values by inverting the sign of the residue if its real part is negative and connecting the entire conductance Y element in an inverted way to the transfer network is indeed new (I think I read pretty much every paper about this topic).

Your method seems to be even better than the method presented in C. Chou and José E. Schutt-Ainé, 2021 for two reasons:

(1) instead of distinguishing between 1 of 12 cases (R < 0, L<0, C<0 for complex poles and R < 0, L < 0 for real poles) your method distinguishes only 1 case: real(residue) < 0. That saves a lot of work in the synthesizer code.

(2) the generated networks in the method of C. Chou and José E. Schutt-Ainé, 2021 (which they call RLCGE) contain one extra controlled voltage source for each pole that has negative R, L or C. This can potentially add a huge number of extra E-Sources and also extra nodes. Your method will just make all R, L and C positive and connect the respective y-element in an inverted way to the transfer network (in your code the _inv nodes). Those _inv nodes are very efficient because you have only one of them for each port and they are re-used for every pole of that port. So you have only 1 extra component for each port and 1 extra node.

**Comment by fnordpole** [comment]
I wrote a new circuit synthesizer that reduces the node and component count growth in the synthesized circuits to O(N) compared to O(N^2) in the current scikit-rf implementation.

For small port counts this may not make a huge difference but for larger port counts it can be significant.

The code is attached. More comments are in the code. I would be happy for comments and feedback.  If you like it, feel free to put the code into scikit-rf. 

```python
def write_spice_subcircuit_new_s(self, file: str, fitted_model_name: str = "s_equivalent") -> None:
	"""
	Creates an equivalent N-port subcircuit based on its vector fitted S parameter responses
	in spice simulator netlist syntax

	Parameters
	----------
	file : str
		Path and filename including file extension (usually .sNp) for the subcircuit file.
	fitted_model_name: str
		Name of the resulting model, default "s_equivalent"

	Returns
	-------
	None

	Notes
	-----
	This new circuit synthesis method reduces the number of voltage or current sources
	in the synthesized circuit from 1+2*N+4*N^2 (current scikit-rf method)
	to 1+6*N (new method)

	The node count for the transfer network is reduced from N+2*N^2 (current
	scikit-rf method) to 3*N (new method)

	The linear growth in component and node count compared to quadratic growth
	can help a lot especially as port counts are getting higher.

	It uses a differential incident wave (a) and a center node on which all positive
	and negative transfer currents are summed. The excess current represents the
	reflected wave (b).

	The approach of having a port network and a transfer network, as described in [2]
	is used in this method. Also, the approach implemented in the current scikit-rf
	circuit synthesis code for dealing with negative component values by multiplying
	the residue by -1 if the real part of it is negative is also used in the new
	method but the negative currents are handled differently.

	I am happy to hear your feedback or comments. I hope there is no bug in the spice
	version. I only tested a four-port network in the spectre simulator and it shows
	the exact same simulated s-parameters as with the current scikit-rf method.

	Best regards,
	Thomas Schweiger
	fnordpole@gmail.com

	Examples
	--------
	Load and fit the `Network`, then export the equivalent subcircuit:

	>>> nw_3port = skrf.Network('my3port.s3p')
	>>> vf = skrf.VectorFitting(nw_3port)
	>>> vf.vector_fit(n_poles_real=1, n_poles_cmplx=4)
	>>> vf.write_spectre_subcircuit_new_s('/my3port_model.sp')

	References
	----------
	.. [1] G. Antonini, "SPICE Equivalent Circuits of Frequency-Domain Responses", IEEE Transactions on
		Electromagnetic Compatibility, vol. 45, no. 3, pp. 502-512, August 2003,
		doi: https://doi.org/10.1109/TEMC.2003.815528

	.. [2] C. -C. Chou and J. E. Schutt-Ainé, "Equivalent Circuit Synthesis of Multiport S Parameters in
		Pole–Residue Form," in IEEE Transactions on Components, Packaging and Manufacturing Technology,
		vol. 11, no. 11, pp. 1971-1979, Nov. 2021, doi: 10.1109/TCPMT.2021.3115113

	.. [3] Romano D, Antonini G, Grossner U, Kovačević-Badstübner I. Circuit synthesis techniques of
		rational models of electromagnetic systems: A tutorial paper. Int J Numer Model. 2019
		doi: https://doi.org/10.1002/jnm.2612

	"""

	# List of subcircuits for the equivalent admittances
	subcircuits = []

	# Provides a unique subcircuit identifier (X1, X2, X3, ...)
	def get_new_subckt_identifier():
		subcircuits.append(f'X{len(subcircuits) + 1}')
		return subcircuits[-1]

	with open(file, 'w') as f:
		# write title line
		f.write('* EQUIVALENT CIRCUIT FOR VECTOR FITTED S-MATRIX\n')
		f.write('* Created using scikit-rf vectorFitting.py\n\n')

		# Define the complete equivalent circuit as a subcircuit with one input node per port
		# those port nodes are labeled p1, p2, p3, ...
		# all ports share a common node for ground reference (node 0)
		str_input_nodes = " ".join(map(lambda x: f'p{x + 1}', range(self.network.nports)))

		f.write(f'.SUBCKT {fitted_model_name} {str_input_nodes}\n')
		for n in range(self.network.nports):
			f.write(f'\n* Port network for port {n + 1}\n')

			# Port reference impedance Z0
			f.write(f'R{n + 1} p{n+1} a{n + 1} {np.real(self.network.z0[0, n])}\n')

			# CCVS implementing the reflected wave b.
			# Also used as current sensor to measure the input current
			f.write(f'H_b_{n + 1} a{n + 1} 0 V_c_{n + 1} 1\n')

			f.write(f'* Differential incident wave a sources for transfer from port {n + 1}\n')

			# CCVS and VCVS driving the transfer admittances with 2*a*sqrt(Z0) = V + I*Z0
			f.write(f'H_p_{n + 1} nt_p_{n + 1} nts_p_{n + 1} H_b_{n + 1} {np.real(self.network.z0[0, n])}\n')
			f.write(f'E_p_{n + 1} nts_p_{n + 1} 0 p{n + 1} 0 1\n')

        		# VCVS driving the transfer admittances with -2*a*sqrt(Z0) = -V - I*Z0
                        f.write(f'E_n_{n + 1} 0 nt_n_{n + 1} nt_p_{n + 1} 0 1\n')

			f.write(f'* Center net for transfer to port {n + 1}\n')

			# Current sensor for the transfer to current port
			f.write(f'V_c_{n + 1} nt_c_{n + 1} 0 0\n')

			for j in range(self.network.nports):
				f.write(f'* Transfer network from port {j + 1} to port {n + 1}\n')

				# Stacking order in VectorFitting class variables:
				# s11, s12, s13, ..., s21, s22, s23, ...
				i_response = n * self.network.nports + j

				# Start with proportional and constant term of the model
				# H(s) = d + s * e  model
				# Y(s) = G + s * C  equivalent admittance
				g = self.constant_coeff[i_response]
				c = self.proportional_coeff[i_response]

				# R for constant term
				if g < 0:
					f.write(f'R{n + 1}{j + 1} nt_n_{j + 1} nt_c_{n + 1} {np.abs(1 / g)}\n')
				elif g > 0:
					f.write(f'R{n + 1}{j + 1} nt_p_{j + 1} nt_c_{n + 1} {1 / g}\n')

				# C for proportional term
				if c < 0:
					f.write(f'C{n + 1}{j + 1} nt_n_{j + 1} nt_c_{n + 1} {np.abs(c)}\n')
				elif c > 0:
					f.write(f'C{n + 1}{j + 1} nt_p_{j + 1} nt_c_{n + 1} {c}\n')

				# Transfer admittances represented by poles and residues
				for i_pole in range(len(self.poles)):
					pole = self.poles[i_pole]
					residue = self.residues[i_response, i_pole]
					node = get_new_subckt_identifier()

					if np.real(residue) < 0.0:
						# Multiplication with -1 required, otherwise the values for RLC would be negative.
						# This gets compensated by inverting the transfer current direction for this subcircuit
						# by connecting it to the negative side of the differential incident wave
						residue = -1 * residue
						node += f' nt_n_{j + 1} nt_c_{n + 1}'
					else:
						# Otherwise it is connected to the positive side of the differential incident wave
						node += f' nt_p_{j + 1} nt_c_{n + 1}'

					if np.imag(pole) == 0.0:
						# Real pole; Add rl_admittance
						l = 1 / np.real(residue)
						r = -1 * np.real(pole) / np.real(residue)
						f.write(node + f' rl_admittance res={r} ind={l}\n')
					else:
						# Complex pole of a conjugate pair; Add rcl_vccs_admittance
						l = 1 / (2 * np.real(residue))
						b = -2 * (np.real(residue) * np.real(pole) + np.imag(residue) * np.imag(pole))
						r = -1 * np.real(pole) / np.real(residue)
						c = 2 * np.real(residue) / (np.abs(pole) ** 2)
						gm = b * l * c
						f.write(node + f' rcl_vccs_admittance res={r} cap={c} ind={l} gm={gm}\n')

		f.write('.ENDS {fitted_model_name}\n\n')

		# Subcircuit for an RLCG equivalent admittance of a complex-conjugate pole-residue pair. Details in [1]
		f.write('.SUBCKT rcl_vccs_admittance n_pos n_neg res=1e3 cap=1e-9 ind=100e-12 gm=1e-3\n')
		f.write('L1 n_pos 1 {ind}\n')
		f.write('C1 1 2 {cap}\n')
		f.write('R1 2 n_neg {res}\n')
		f.write('G1 n_pos n_neg 1 2 {gm}\n')
		f.write('.ENDS rcl_vccs_admittance\n\n')

		# Subcircuit for an RL equivalent admittance of a real pole-residue pair. Details in [1]
		f.write('.SUBCKT rl_admittance n_pos n_neg res=1e3 ind=100e-12\n')
		f.write('L1 n_pos 1 {ind}\n')
		f.write('R1 1 n_neg {res}\n')
		f.write('.ENDS rl_admittance\n\n')
```

**Comment by Vinc0110** [comment]
Thanks for all the investigations, analyses, and comparisons on the topic! Very interesting! 

It's funny that I apparently worked on the circuit synthesis at the same time (year 2020 more or less) like Chou and Schutt-Ainé and came to similar results. I did not think it would be paper material, because similar stuff existed in ADS and other simulators for years... But I'm glad my method compares well against theirs :sunglasses: It would be super interesting to see the commercial solutions, but they are closed source, of course.

Your new synthesizer code sounds like a great improvement, but I still need to read and understand it properly. I suggest you create a pull request here on github. It would also make sense to split the discussions of the code from this thread, which is already very long.

In case you're not familiar with the process, we have a [short guide in the docs](https://scikit-rf.readthedocs.io/en/latest/contributing/index.html#basic-git-command-line-workflow) and there are millions of tutorials online. Let me know if you need any help.

**Reply by fnordpole** [reply]
Pull request is done. Would you be motivated to co-author a paper with me about the new synthesizer? From my part I would describe the differential incident wave topology with the current summing node reducing the node-count and component-count to linear and from your part we could put the negative-impedance handling in. Both in combination is a pretty powerful approach. Also the paper would serve as nice documentation for the code. Maybe same journal (ieee) as the one Chou and Schutt-Ainé published, as a follow up to their paper.

**Reply by Vinc0110** [reply]
Thanks for the pull request. This is a busy time for me at work, but I'll try to review it as soon as possible.

Writing a paper about it sounds like fun. We can give it a shot. Thanks for asking! We should probably discuss the details somewhere else. 
[Email address removed]
